zebra-network/src/peer_set/inventory_registry.rs

```diff
@@ -111,9 +215,14 @@ impl InventoryRegistry {
         Ok(())
     }
 
-    /// Record that the given inventory `hash` is available from the peer `addr`
-    fn register(&mut self, hash: InventoryHash, addr: SocketAddr) {
-        self.current.entry(hash).or_default().insert(addr);
+    /// Record the given inventory `change` for the peer `addr`.
+    fn register(&mut self, change: InventoryChange) {
+        let status = change.as_ref().map(|_| ());
+        let (invs, addr) = change.inner();
+
+        for inv in invs {
+            self.current.entry(inv).or_default().insert(addr, status);
```

Comment:
TODO: should we prefer existing `notfound` entries to new `inv` entries?

Reply:
I'm not sure, but my first thought would be no. I think we should keep the latest information, so the current code seems correct because it overwrites any existing entries for the peer.

Reply:
I thought so too, but then I realised that a malicious peer could repeatedly `notfound` a request, then advertise the same inventory. That would funnel a lot of failing requests to that peer.

So I think we should prefer `missing` over `advertised`, and prefer the current period over the previous one. (Registry rotation will clear out any `missing` entries every 75 seconds.)

See 05cdda5f9.